[8.0] [l10n_aeat_sii] Adaptaciones para SII versión 1.1

jaumeplanas opened the pull request:
PR para recoger diversas adaptaciones de SII para la versión 1.1:

- [X] Cambio de nombres de etiquetas.

- [X] Macrodato, para informar facturas iguales o superiores a 100 000 euros

- [X] Cambiar el número de version del XML generado a 1.1.

- [X] Cambiar los URIs de los archivos WSDL de la AEAT.

Toda aportación por parte de la comunidad es bienvenida.

Commit:
Adaptations to SII version 1.1

pedrobaeza commented on .gitignore:
```diff
@@ -54,3 +54,5 @@ docs/_build/
 # Backup files
 *~
 *.swp
+/openerp-server.conf
```
Esto no debería añadirse

pedrobaeza commented on l10n_es_aeat_sii/models/account_invoice.py:
```diff
@@ -46,14 +47,15 @@ def empty_decorator_factory(*argv, **kwargs):
-SII_VERSION = '1.0'
-SII_START_DATE = '2017-07-01'
+SII_VERSION = '1.1'
+SII_START_DATE = '2018-07-01'
```
Esta fecha no debería cambiarse, ya que fue para el envío del primer semestre de 2017

pedrobaeza commented on l10n_es_aeat_sii/models/account_invoice.py:
```diff
@@ -153,6 +165,11 @@ def _default_sii_registration_key(self):
     sii_property_cadastrial_code = fields.Char(
         string="Real property cadastrial code", size=25, copy=False,
     )
+    sii_macrodata = fields.Boolean(
```
Esto debería hacerse como un campo computed, en lugar de sobreescribir el método _compute_amount, hacer super registro a registro (desperdiciando posibles mejoras de proceso por lotes), y haciendo un write dentro de un compute.

pedrobaeza commented on l10n_es_aeat_sii/models/account_invoice.py:
```diff
@@ -461,10 +478,13 @@ def _get_sii_out_taxes(self):
                     ex_taxes = taxes_sfesbe
                     if tax_line in ex_taxes:
-                        sub_dict.setdefault('Exenta', {'BaseImponible': 0})
+                        det_dict = sub_dict.setdefault('Exenta', {})
```
Haz la creación de esto en un solo paso por claridad de lectura.

pedrobaeza commented on l10n_es_aeat_sii/models/account_invoice.py:
```diff
@@ -504,13 +524,16 @@ def _get_sii_out_taxes(self):
                     if tax_line in taxes_sfesse:
-                        exempt_dict = service_dict['Sujeta'].setdefault(
-                            'Exenta', {'BaseImponible': 0},
+                        exempt_dict = service_dict.setdefault('Exenta', {})
```
Haz la creación de esto en un solo paso por claridad de lectura.

pedrobaeza commented on l10n_es_aeat_sii/models/aeat_sii_map.py:
```diff
@@ -10,34 +10,34 @@ class AeatSiiMap(models.Model):
-            raise exceptions.Warning(
-                _("Error! The dates of the record overlap with an existing "
-                  "record."))
+        for rec in self:
```
Por favor no generes este diff tan grande por cambiar el `@api.one`. Si quieres hacerlo, deja el método tal cual, quitando los decoradores, y creas otro método ya api.multi que llame a éste.

pedrobaeza commented on l10n_es_aeat_sii/tests/test_l10n_es_aeat_sii.py:
```diff
@@ -142,10 +144,10 @@ def _get_invoices_test(self, invoice_type, special_regime):
                 },
                 'DescripcionOperacion': u'/',
-                'ClaveRegimenEspecialOTrascendencia': special_regime,
+                'ClaveRegimenEspecialOTrascendencia': unicode(special_regime),
```
Por qué unicode si son caracteres normales?

jaumeplanas replied:
Porque, para este valor, el método invoice._get_sii_invoice_dict() devuelve un valor unicode, por lo que el método _get_invoices_test() debe crear también un valor unicode para que el método assertDictEqual() no detecte que son diccionarios distintos y genere un falso error.

pedrobaeza replied:
Pero aunque sea unicode, si no hay caracteres raros, la igualdad se cumple, y esto estaba funcionando antes AFAIK

jaumeplanas replied:
Este cambio lo realicé después de analizar el primer log de errores de Travis, que decía que los dos diccionarios eran distintos. Pero si lo consideras oportuno, lo dejo tal como estaba antes.

pedrobaeza replied:
A ver, lo que me extraña es que esta parte en teoría no ha cambiado, y antes daba correcto. ¿Por qué ahora no?

pedrobaeza commented on l10n_es_aeat_sii/tests/test_l10n_es_aeat_sii.py:
```diff
@@ -215,6 +218,7 @@ def test_get_invoice_data(self):
         self.invoice.supplier_invoice_number = 'sup0001'
         invoices = self.invoice._get_sii_invoice_dict()
         test_in_invoice = self._get_invoices_test('F1', '01')
+        print invoices, test_in_invoice
```
Nunca incluyas `print`

pedrobaeza commented on l10n_es_aeat_sii/tests/test_l10n_es_aeat_sii.py:
```diff
@@ -266,6 +270,6 @@ def test_sii_description(self):
-    def test_permissions(self):
-        """This should work without errors"""
-        self.invoice.sudo(self.user).signal_workflow('invoice_open')
+    # def test_permissions(self):
```
Por qué comentas este test?

pedrobaeza commented on l10n_es_aeat_sii/views/account_invoice_view.xml:
```diff
@@ -106,6 +107,8 @@
                                    attrs="{'invisible': [('sii_registration_key_code', 'not in', ['12', '13'])],
                                            'required': [('sii_registration_key_code', 'in', ['12', '13']), ('sii_property_location', 'in', ['1', '2'])]}"/>
+                            <field name="sii_macrodata"/>
+
```
No añadas líneas en blanco innecesarias

pedrobaeza reviewed (changes requested):
Aparte de los comentarios inline, la mayor duda de esto es: ¿hay que esperarse a la fecha de entrada en vigor para poder poner esto, o desde ya se podría poner?

jabibi commented:
Buenas. Estos mismos cambios son aplicables a la versión 10.0?
Sabéis si se puede utilizar ya en producción la versión 1.1?

Me voy de vaciones del 23 de junio al 8 de julio y justo el cambio de versión pilla en medio

jabibi commented on l10n_es_aeat_sii/models/account_invoice.py:
```diff
@@ -46,14 +47,15 @@ def empty_decorator_factory(*argv, **kwargs):
     'MQ': 'FR',
     'GF': 'FR',
 }
+SII_MACRODATA_LIMIT = 100000.0
```
Según el documento `2018 - Resumen mejoras SII 2018.pdf` aportado por Gafic el límite sería 100.000.000€:

> Macrodato:
> El 11-05-2018 este umbral cambió de 100.000 a 100.000.000€.
> Se crea la etiqueta <Macrodato> en los libros de emitidas y recibidas para
> identificar aquellas facturas con importe de la factura superior a un umbral de
> 100.000.000 euros. Viene a ser una confirmación de que la factura informada, de
> gran valor, es correcta.

pedrorgil commented on l10n_es_aeat_sii/models/account_invoice.py:
```diff
@@ -461,10 +478,13 @@ def _get_sii_out_taxes(self):
-                        sub_dict.setdefault('Exenta', {'BaseImponible': 0})
+                        det_dict = sub_dict.setdefault('Exenta', {})
+                        det_dict.setdefault('DetalleExenta',
+                                            {'BaseImponible': 0})
```
DetalleExenta debe ser una lista de diccionarios de la forma [{'BaseImponible': X, 'CausaExencion': Y},...], no un único diccionario

pedrorgil commented on l10n_es_aeat_sii/models/account_invoice.py:
```diff
@@ -461,10 +478,13 @@ def _get_sii_out_taxes(self):
-                            sub_dict['Exenta']['CausaExencion'] = exempt_cause
-                        sub_dict['Exenta']['BaseImponible'] += (
+                            det_dict['DetalleExenta'][
+                                'CausaExencion'] = exempt_cause
```
Igual aquí

pedrorgil commented on l10n_es_aeat_sii/models/account_invoice.py:
```diff
@@ -461,10 +478,13 @@ def _get_sii_out_taxes(self):
+                            det_dict['DetalleExenta'][
+                                'CausaExencion'] = exempt_cause
+                        det_dict['DetalleExenta']['BaseImponible'] += (
                             inv_line._get_sii_line_price_subtotal()
```
Aquí no hay que sumar, debe ser la base imponible de cada línea por separado en la lista de diccionarios

pedrorgil commented on l10n_es_aeat_sii/models/account_invoice.py:
```diff
@@ -504,13 +524,16 @@ def _get_sii_out_taxes(self):
+                            exempt_dict['DetalleExenta'][
+                                'CausaExencion'] = exempt_cause
+                        exempt_dict['DetalleExenta']['BaseImponible'] += (
+                            inv_line._get_sii_line_price_subtotal())
```
Todo este bloque lo mismo, es una lista de diccionarios

jaumeplanas commented:
Respecto a la fecha de entrada en vigor de la versión 1.1, tal como está codificado, no permite tener diferentes versiones simultáneamente. Creo que deberá actualizarse este módulo a partir del 1 de julio de 2018

pedrobaeza commented:
Uf, eso no es demasiado operativo, la verdad, ya que obliga a ponerlo justo en esa fecha. Deberías poner la versión de SII condicional según la fecha, y luego leer esa versión para poner una clave u otra.

jaumeplanas commented:
OK, ahora modifico la versión en función de la fecha, aunque solo recordar que sí será necesario cambiar en esta fecha concreta los parámetros de sistema con los endpoints WDSL, que han cambiado para la versión 1.1.

pedrobaeza commented:
Joer, es que vaya Hacienda las cosas malas que hacen...

Sí, lo de los endpoint va a ser también un problema. En v9+ podemos manejarlo como digo creando 2 agencias, y cambiando ese mismo día la agencia implicada. En v8 habrá que hacerlo manual, o hacer un backport de esa característica.

pedrobaeza commented on l10n_es_aeat_sii/models/aeat_sii_map.py:
```diff
@@ -36,8 +39,8 @@ def _unique_date_range(self):
             raise exceptions.Warning(
-                _("Error! The dates of the record overlap with an existing "
-                  "record."))
+                _("Error! The dates of the record overlap with an existing"
```
Por qué hacer este cambio si estaba bien antes? Siempre la máxima debe ser reducir el diff para mejor legibilidad

jabibi commented:
Por lo que me comentan desde otra empresa, hasta el 15 de julio se puede seguir utilizando la versión 1.0
Estoy buscando una fuente oficial sobre ese dato pero no lo encuentro

pedrobaeza commented on l10n_es_aeat_sii/views/account_invoice_view.xml:
```diff
@@ -59,13 +58,11 @@
                                    readonly="1"
-                                   context="{'tree_view_ref': 'l10n_es_aeat_sii.view_queue_job_sii'}"
-                            />
+                                   context="{'tree_view_ref': 'l10n_es_aeat_sii.view_queue_job_sii'}"/>
```
Este cambio no es necesario y de hecho es mejor la otra forma, ya que admite añadir atributos sin tener 2 líneas de diff

pedrobaeza commented on l10n_es_aeat_sii/views/account_invoice_view.xml:
```diff
@@ -127,13 +125,11 @@
                                    readonly="1"
-                                   context="{'tree_view_ref': 'l10n_es_aeat_sii.view_queue_job_sii'}"
-                            />
+                                   context="{'tree_view_ref': 'l10n_es_aeat_sii.view_queue_job_sii'}"/>
```
Este cambio no es necesario y de hecho es mejor la otra forma, ya que admite añadir atributos sin tener 2 líneas de diff

pedrobaeza commented on l10n_es_aeat_sii/models/account_invoice.py:
```diff
@@ -449,7 +472,7 @@ def _get_sii_out_taxes(self):
             for tax_line in inv_line.invoice_line_tax_id:
                 breakdown_taxes = (
-                    taxes_sfesb + taxes_sfesisp + taxes_sfens + taxes_sfesbe
+                        taxes_sfesb + taxes_sfesisp + taxes_sfens + taxes_sfesbe
```
Por qué esta indentación extra?

pedrobaeza commented on l10n_es_aeat_sii/models/account_invoice.py:
```diff
@@ -417,9 +440,9 @@ def _sii_adjust_first_semester(self, taxes_dict):
-                                },
                             },
-                        )
+                        },
```
No entiendo estos cambios en la indentación

pedrobaeza reviewed (changes requested):
@jaumeplanas veo que estás haciendo un montón de cambios en estilo e indentaciones que, aparte de incorrectos, no son necesarios, ya que aumenta el diff sin ningún beneficio.

Aparte, lo de la versión sigue sin ser completo, ya que lo de `PeriodoLiquidacion` sigue hardcodeado sin revisar qué version es la que se está usando.

ELNOGAL commented:
Hola:
En preguntas técnicas frecuentes de la AEAT dicen que existirá un periodo de convivencia hasta el 20 de julio. 
Pongo enlace:
[https://www.agenciatributaria.es/AEAT.internet/Inicio/Ayuda/Modelos__Procedimientos_y_Servicios/Ayuda_P_G417____IVA__Llevanza_de_libros_registro__SII_/Ayuda_tecnica/Informacion_tecnica_SII/Preguntas_tecnicas_frecuentes/2__Cuestiones_Version_SII_1_1/3___Va_a_existir_un_periodo_de_convivencia_de_las_versiones_1_0_y_1_1__.shtml](url)

jabibi commented:
Gracias @ELNOGAL, eso es justo lo que andaba buscando

jabibi commented:
@jaumeplanas @pedrobaeza con el tema de la versión yo no me volvería loco. Es solo cambiar el valor de SII_VERSION y empezar a usar el nuevo módulo a partir del 1 de julio. De esta forma no hace falta cambiar SII_START_DATE ni añadir condiciones a la versión

jaumeplanas commented:
@pedrobaeza Respecto al tema de las indentaciones, ha sido mi configuración de PyCharm, que reformatea el código automáticamente antes de realizar un commit. Ya he desactivado esta función. Lo siento.
Respecto al PeriodoLiquidacion, creo que no está relacionado con la versión de SII, sino que recoge información del ejercicio fiscal y periodo de la factura.

pedrobaeza commented:
@jabibi pero el problema es que aunque hay la flexibilidad de plazo, no debería dejarse esto a una acción manual de sustitución del código. No es demasiado jaleo tampoco creo yo.

pedrobaeza commented:
@jaumeplanas la clave ha cambiado de nombre en esta versión, antes `PeriodoImpositivo`, y ahora `PeriodoLiquidacion`, así que sí que tiene que ver con la versión. Lo ponia como ejemplo de las cosas que habría que poner condicionales.

jaumeplanas commented:
@pedrobaeza Es cierto, hay cambio de etiquetas en función de la versión. Pero no compartimos tu opinión de condicionar toda la lógica de creación del diccionario global a una versión concreta. Si seguimos en esta dirección, habrá que poner condicionales para incluir o no el tag `DetalleExenta`, por ejemplo, que hace que la construcción del diccionario sea muy distinta, con lo que llenaremos de `if`s todo el código.
Además, cuando publiquen una nueva versión 1.2, por ejemplo, ¿seguiremos añadiendo nuevas condiciones y complicando el código con condicionales de versiones que ya no se utilizarán más?

pedrobaeza commented:
Estoy de acuerdo que complica, pero como dicen en mi tierra, "las cosas bien hechas, bien parecen". Entonces si no ponemos la lógica condicional, esto no lo fusionaré hasta el mismo día de entrada en vigor de la versión. Al final, hay 20 días de margen con el sistema antiguo.

jabibi commented:
Es lo que os decía y lo que comenta @jaumeplanas. Yo no me complicaría la vida con esto. Se hace con arreglo a la última versión y se empieza a utilizar el módulo modificado cuando toque

dbardelas commented:
La etiqueta de Macrodato, según la información técnica de los servicios Web de la AEAT, es para informar facturas iguales o superiores a 100.000.000 €. En el primer comentario, pone 100.000 €.

omar7r commented on l10n_es_aeat_sii/models/account_invoice.py:
```diff
@@ -153,11 +159,26 @@ def _default_sii_registration_key(self):
+        string="MacroData",
+        help="Check to confirm that the invoice has an absolute amount "
+             "greater o equal to 100 000 000,00 euros.",
+        compute='_compute_macrodata', store=True
```
Para esto estaría bien un pequeño premigration ya que con una query se calcula muy rápido el valor inicial, es que sino tarda muchísimo en hacer el update la primera vez del módulo para una base de datos que tenga muchas facturas.

omar7r replied:
Por otra parte, faltaría actualizar la versión del módulo

omar7r replied:
Por lo demás a nivel funcional, funciona bien

enriquemartin committed:
[FIX] l10n_es_aeat_sii - corrección en DesgloseTipoOperacion/PrestacionServicios/Sujeta: si el impuesto es por servicios, sujeto y no exento el diccionario se generaba mal --> error

enriquemartin commented:
Hola!
Hemos detectado un problema en facturas expedidas, desglose tipo operación, prestaciones de servicios (las relativas a 21% IVA (Servicios)).
Hay un PR en https://github.com/minorisa/l10n-spain/pull/1

omar7r committed:
[FIX] l10n_es_aeat_sii: Se corrige el envío de factuas con más de una linea exenta. [REF] l10n_es_aeat_sii: Se eliminan las tabulaciones extra, respetando PEP8

omar7r commented:
Hola,

Se ha encontrado un bug al enviar facturas con varias lineas exentas, o todas en el caso de una entrega intracomunitaria, crea una causa por linea de factura causando que falle el SII si hay más de 7 lineas, en lugar de agrupar por causa. Se ha hecho un PR contra este en https://github.com/minorisa/l10n-spain/pull/2

Un saludo

JuanjoA commented:
Buenas, llego tarde, ¿en qué puedo ayudar?, de momento testearé la v8.

jabibi reviewed (changes requested):
Se me hace imposible seguir los cambios de esta PR. Por qué salen otros módulos? es de un rebase o algo así? aun así, si esos cambios están en la rama principal no deberían salir en el diff

omar7r commented:
Hay que hace rebase, mi rama desde donde les propuse mi PR estaba más actualizada que esta

pedrobaeza committed:
[FIX] l10n_es_aeat_sii: Don't store macrodata compute

This is only needed when sending to SII, and this will make the installation/update
very slow.

Abranes reviewed (approved):
Lo hemos estado probando y quitado de lo que tarda a instalar por el compute del macrodata (que acaba de corregir @pedrobaeza) todo funcionando a la perfección.

mlopezgaldon commented on l10n_es_aeat_sii/models/account_invoice.py:
```diff
@@ -862,18 +900,21 @@ def _get_sii_invoice_dict_in(self, cancel=False):
+                                                         2), 2) or 0.0),
             }
+            if self.sii_macrodata:
+                inv_dict["FacturaExpedida"].update(Macrodato="S")
```
inv_dict["FacturaRecibida"].update(Macrodato="S")

Esto se debió copiar y pegar desde facturas expedidas, esta función es para recibidas.

(No sé si el método revisión es el apropiado para mí, pero considero que se trata de un typo demasiado pequeño como para forkear)

pedrobaeza replied:
Vaya, si lo hubieras dicho antes de fusionar... Bueno, hago commit directo para corregirlo. Además, no creo que se dé el caso de alguien que tenga esta factura durante estos minutos, jeje.

pedrobaeza replied:
Corregido en 3e1a26ddb07fba31533d31ff8f81dd182469bc41

pedrobaeza reviewed (approved):
Bueno, aunque el diff es un follón por cambios estilísticos que no deberían haberse hecho, no merece la pena invertir más tiempo en esta versión, así que fusiono, habiendo quitado que no se almacene lo del macrodato para no afectar el rendimiento de la actualización

pedrobaeza commented:
Gracias a todos los colaboradores y testers.
